api: fix nmd_get_next_frame() when reaching EOF

mbouron opened the pull request:
Probably incomplete.

ubitux commented on src/api.c:
```diff
@@ -591,8 +592,10 @@ static AVFrame *pop_frame(struct nmd_ctx *s)
-            if (ret < 0)
+            if (ret < 0) {
                 TRACE(s, "poped a message raising %s", av_err2str(ret));
+                s->eof = (ret == AVERROR_EOF);
```
Is this really enough? If there is a trim duration specified, I think the filtering mod will early exit with an `AVERROR_EXIT`

mbouron replied:
Good catch, I double checked and added AVERROR_EXIT.

ubitux reviewed (commented):
So previously when we reached EOF (or early exit), be it through get next frame or a traditional time based query, the threads were stopped, releasing a bunch of resources. This is not the case anymore?

mbouron committed:
tests: add audio_start_end_time test

mbouron commented:
> So previously when we reached EOF (or early exit), be it through get next frame or a traditional time based query, the threads were stopped, releasing a bunch of resources. This is not the case anymore?

Yes, this is not the case anymore as it was causing issues with VAAPI.

ubitux commented on tests/test_audio_start_end_time.c:
```diff
@@ -0,0 +1,54 @@
+    nmd_set_option(s, "use_pkt_duration", use_pkt_duration);
+    nmd_set_option(s, "avselect", NMD_SELECT_AUDIO);
+    nmd_set_option(s, "audio_texture", 0);
+    nmd_set_option(s, "end_time", 30.0);
+    nmd_set_option(s, "end_time", 90.0);
```
Why the dup? Did you meant `start` for the first one?

mbouron replied:
Fixed.

ubitux reviewed (commented):
I think the bump to FFmpeg 6 belongs in a separate PR

ubitux commented on src/api.c:
```diff
@@ -886,6 +889,23 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
+        s->last_pushed_frame_ts = AV_NOPTS_VALUE;
+
+        const struct nmdi_opts *o = &s->opts;
+        nmdi_async_seek(s->actx, o->start_time64);
```
Missing return check

mbouron replied:
Fixed.

ubitux commented on src/api.c:
```diff
@@ -886,6 +889,23 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
+        const struct nmdi_opts *o = &s->opts;
+        nmdi_async_seek(s->actx, o->start_time64);
+
+        return ret_frame(s, NULL);
```
this is redundant, `frame` is `NULL` so you can rely on the common path

mbouron replied:
Fixed.

ubitux commented on src/api.c:
```diff
@@ -886,6 +889,23 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
+         * return, we need to seek back to the beginning of the file because
+         * nmd_get_next_frame() is expected to loop around when reaching EOF.
+         */
+        s->eof = 0;
+
```
So this is going to cause a (costly) seek systematically at the end. Shouldn't that happen only if get_next_frame is called again after a `NULL` was returned?

mbouron replied:
It depends on the scenario, this can have an advantage: sending an seek command asynchronously at the end will make the next get_next_frame() faster.

ubitux replied:
Right but the classic scenario of `get_next_frame()` is to go through all the frame and stop at the end. This scenario get hurt a lot and the user has no way of avoiding it. And in the scenario where the user wants the loop, the seek must be done anyway, so I think that's fine to have it lazily.

mbouron replied:
Done.

ubitux commented on src/api.c:
```diff
@@ -885,7 +889,29 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
+    }
+
     AVFrame *frame = pop_frame(s);
+    if (!frame && s->eof) {
+        /*
+         * If the async module reached EOF and there is no more frame to
+         * return, we need to seek back to the beginning of the file because
+         * nmd_get_next_frame() is expected to loop around when reaching EOF.
+         */
+        s->eof = 0;
+        s->loop = 1;
+    }
```
It doesn't work if you remove this block and instead of checking of `s->loop` you check for `s->eof` and the next iteration?

```c
if (s->eof) {
    // seek
}
AVframe *frame = pop_frame(s);
return ret_frame(s, frame);
```

Am I missing something?

mbouron replied:
Checking for eof this way is not enough, you need to actually have reached eof and have no more frame to return.

ubitux replied:
But if the previous call set `s->eof`, it also necessarily returned a `NULL` frame?

ubitux replied:
Ah mmh ok I think I get it, my bad.

mbouron committed:
api: allow pop_frame() to forward errors to its caller

mbouron committed:
api: add status argument to ret_frame()

mbouron committed:
api: fix nmd_get_next_frame() when reaching EOF

Before faa89639f4d1fa071a6daa9acb0b5718fab34667, nmd_get_next_frame()
relied on the behaviour of nmdi_async_pop_frame() which stopped the
whole async module at EOF. Since this is not the case anymore, we need
to seek back to the beginning on the file when nmd_get_next_frame()
reaches EOF as it is supposed to loop around in this situation.

Fixes a regression introduced by
faa89639f4d1fa071a6daa9acb0b5718fab34667.

ubitux commented on src/api.c:
```diff
@@ -61,6 +61,7 @@ struct nmd_ctx {
     int64_t last_frame_poped_ts;
     int64_t first_ts;
     int64_t last_ts;
+    int loop;
```
I think I actually prefer `eof` with the following comment: "set if the latest frame returned was NULL and meant EOF"

mbouron replied:
Done.

ubitux commented on src/api.c:
```diff
@@ -469,7 +470,17 @@ static struct nmd_frame *ret_frame(struct nmd_ctx *s, AVFrame *frame, int status
     struct nmd_frame *ret = NULL;
     const struct nmdi_opts *o = &s->opts;
 
+    if (status == 0)
+        s->loop = 0;
+
     if (!frame) {
+        /*
+         * If the async module reached EOF and there is no more frame to
+         * return, we need to seek back to the beginning of the file because
+         * nmd_get_next_frame() is expected to loop around when reaching EOF.
+         */
+        if (status == AVERROR_EOF || status == AVERROR_EXIT)
+            s->loop = 1;
```
Why not simply set `s->eof = !frame && (status == AVERROR_EOF || status == AVERROR_EXIT)` unconditionally?

mbouron replied:
Done.

ubitux commented on src/api.c:
```diff
@@ -699,6 +710,12 @@ struct nmd_frame *nmd_get_frame_ms(struct nmd_ctx *s, int64_t t64)
 
     START_FUNC_T("GET FRAME", t64 / 1000000.);
 
+    /*
+     * Clear the loop flag if nmd_get_frame() is used after nmd_get_next_frame()
+     * reached EOF.
+     */
+    s->loop = 0;
+
```
This wouldn't be needed then

ubitux commented on src/api.c:
```diff
@@ -888,6 +905,18 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
         return ret_frame(s, NULL, ret);
 
+    if (s->loop) {
+        s->loop = 0;
```
this assignment neither

mbouron replied:
Done.

ubitux reviewed (approved):
LGTM thanks a lot :)
